### PR TITLE
[Doc] Update doc of training and inference output.

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/source/advanced/multi-task-learning.rst [hunk 1]
```diff
@@ -318,3 +318,129 @@ GraphStorm supports to run multi-task inference on :ref:`SageMaker<distributed-s
         --instance-count <INSTANCE_COUNT> \
         --instance-type <INSTANCE_TYPE>
 
+Multi-task Learning Output
+--------------------------
+
+Saved Node Embeddings
+~~~~~~~~~~~~~~~~~~~~~~
+When ``save_embed_path`` is provided in the training config or inference condig,
```
Comment (zhjwy9343) [inline docs/source/advanced/multi-task-learning.rst:326]:
```suggestion
When ``save_embed_path`` is provided in the training configuration or inference configuration YAML file,
```

File: docs/source/advanced/multi-task-learning.rst [hunk 2]
```diff
@@ -318,3 +318,129 @@ GraphStorm supports to run multi-task inference on :ref:`SageMaker<distributed-s
...
+GraphStorm will save the node embeddings in the corresponding path.
+In multi-task learning, by default, GraphStorm will save the node embeddings
+produced by the GNN layer for every node type under the path specified by
+``save_embed_path``。 The output format follows the :ref:`GraphStorm saved node embeddings
```
Comment (zhjwy9343) [inline docs/source/advanced/multi-task-learning.rst:330]:
```suggestion
``save_embed_path``. The output format follows the :ref:`GraphStorm saved node embeddings
```

File: docs/source/advanced/multi-task-learning.rst [hunk 3]
```diff
@@ -318,3 +318,129 @@ GraphStorm supports to run multi-task inference on :ref:`SageMaker<distributed-s
...
+format<gs-out-embs>`. Meanwhile, in multi-task learning, certain tasks might apply
+task specific normalization to node embeddings. For instance, a link prediction
+task might apply l2 normalization on each node embeddings. In certain cases, GraphStorm
+will also save the normalized node embeddings under ``save_embed_path``.
+The task specific node embeddings are saved separately under different sub-directories
+named with the corresponding task id. (A task id is formated as ``<task_type>-<ntype/etype(s)>-<label>``.
+For instance, the task id of a node classification task on the node type ``paper`` with the
+label filed ``venue`` will be ``node_classification-paper-venue``. As another example,
```
Comment (zhjwy9343) [inline docs/source/advanced/multi-task-learning.rst:338]:
```suggestion
label field ``venue`` will be ``node_classification-paper-venue``. As another example,
```

File: docs/source/advanced/multi-task-learning.rst [hunk 4]
```diff
@@ -318,3 +318,129 @@ GraphStorm supports to run multi-task inference on :ref:`SageMaker<distributed-s
...
+the task id of a link prediction task on the edge type ``(paper, cite, paper)`` will be
+``link_prediction-paper_cite_paper``
+and the task id of a edge regression task on the edge type ``(paper, cite, paper)`` with
+the label field ``year`` will be ``edge_regression-paper_cite_paper-year``).
+The output format of task specific node embeddings follows
+the :ref:`GraphStorm saved node embeddings format<gs-out-embs>`.
+The ``save_embed_path`` in multi-task learning will look like following:
```
Comment (zhjwy9343) [inline docs/source/advanced/multi-task-learning.rst:345]:
```suggestion
The contents of  the ``save_embed_path`` in multi-task learning will look like following:
```

File: docs/source/advanced/multi-task-learning.rst [hunk 5]
```diff
@@ -318,3 +318,129 @@ GraphStorm supports to run multi-task inference on :ref:`SageMaker<distributed-s
...
+            ntype1/
+                embed_nids-00000.pt
+                embed_nids-00001.pt
+                ...
+                embed-00000.pt
+                embed-00001.pt
+                ...
+            emb_info.json
+
+In the above example both the link prediction task and the edge regression
```
Comment (zhjwy9343) [inline docs/source/advanced/multi-task-learning.rst:398]:
```suggestion
In the above example both the link prediction and the edge regression task
```

File: docs/source/advanced/multi-task-learning.rst [hunk 6]
```diff
@@ -318,3 +318,129 @@ GraphStorm supports to run multi-task inference on :ref:`SageMaker<distributed-s
...
+apply task specific normalization on node embeddings.
+
+**Note: The built-in GraphStorm training or inference pipeline
+(launched by GraphStorm CLI) will process each saved node embeddings
```
Comment (zhjwy9343) [inline docs/source/advanced/multi-task-learning.rst:402]:
```suggestion
(launched by GraphStorm CLIs) will process each saved node embeddings
```

File: docs/source/advanced/multi-task-learning.rst [hunk 7]
```diff
@@ -318,3 +318,129 @@ GraphStorm supports to run multi-task inference on :ref:`SageMaker<distributed-s
...
+to convert the integer node ids into the raw node ids, which are usually string node ids..**
```
Comment (zhjwy9343) [inline docs/source/advanced/multi-task-learning.rst:403]:
```suggestion
to convert the integer node IDs into the raw node IDs, which are usually string IDs.**
```

File: docs/source/advanced/multi-task-learning.rst [hunk 8]
```diff
@@ -318,3 +318,129 @@ GraphStorm supports to run multi-task inference on :ref:`SageMaker<distributed-s
...
+Details can be found in :ref:`GraphStorm Output Node ID Remapping<output-remapping>`
+
+Saved Prediction Results
+~~~~~~~~~~~~~~~~~~~~~~~~~
+When ``save_prediction_path`` is provided in the inference condig,
```
Comment (zhjwy9343) [inline docs/source/advanced/multi-task-learning.rst:408]:
```suggestion
When ``save_prediction_path`` is provided in the inference configuration YAML file,
```

File: docs/source/advanced/multi-task-learning.rst [hunk 9]
```diff
@@ -318,3 +318,129 @@ GraphStorm supports to run multi-task inference on :ref:`SageMaker<distributed-s
...
+GraphStorm will save the prediction results in the corresponding path.
+In multi-task learning inference, each prediction task will have its prediction
+results saved separately under different sub-directories
+named with the
+corresponding task id. The output format of task specific prediction results
+follows the :ref:`GraphStorm saved prediction result format<gs-out-predictions>`.
+The ``save_prediction_path`` in multi-task learning will look like following:
```
Comment (zhjwy9343) [inline docs/source/advanced/multi-task-learning.rst:415]:
```suggestion
The contents of the ``save_prediction_path`` in multi-task learning will look like following:
```

File: docs/source/advanced/multi-task-learning.rst [hunk 10]
```diff
@@ -318,3 +318,129 @@ GraphStorm supports to run multi-task inference on :ref:`SageMaker<distributed-s
...
+                predict-00001.pt
+                ...
+                predict_nids-00000.pt
+                predict_nids-00001.pt
+                ...
+            result_info.json
+        ...
+
+**Note: The built-in GraphStorm inference pipeline
+(launched by GraphStorm CLI) will process each saved prediction result
```
Comment (zhjwy9343) [inline docs/source/advanced/multi-task-learning.rst:444]:
```suggestion
(launched by GraphStorm CLIs) will process each saved prediction result
```

File: docs/source/advanced/multi-task-learning.rst [hunk 11]
```diff
@@ -318,3 +318,129 @@ GraphStorm supports to run multi-task inference on :ref:`SageMaker<distributed-s
...
+to convert the integer node ids into the raw node ids, which are usually string node ids.**
```
Comment (zhjwy9343) [inline docs/source/advanced/multi-task-learning.rst:445]:
```suggestion
to convert the integer node IDs into the raw node IDs, which are usually string IDs.**
```

File: docs/source/cli/model-training-inference/output.rst [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+**Note: The built-in GraphStorm training or inference pipeline
+(launched by GraphStorm CLI) will process the saved node embeddings
+to convert the integer node ids into the raw node ids, which are usually string node ids..**
+Details can be found in :ref:`GraphStorm Output Node ID Remapping<output-remapping>`
+
+.. _gs-output-predictions:
+
+Saved Prediction Results
+------------------------
+When ``save_prediction_path`` is provided in the inference condig,
```
Comment (zhjwy9343) [inline docs/source/cli/model-training-inference/output.rst:65]:
```suggestion
When ``save_prediction_path`` is provided in the inference configurations,
```

File: docs/source/cli/model-training-inference/output.rst [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+For node prediction tasks, the prediction results are saved per node type.
+GraphStorm will also save an ``result_info.json`` file, which contains all
+the metadata for the saved prediction results. The ``save_prediction_path``
+will look like following:
+
+.. code-block:: bash
+
+    prediction_dir/
+        ntype0/
+            predict-00000.pt
```
Comment (zhjwy9343) [inline docs/source/cli/model-training-inference/output.rst:76]: ditto

File: docs/source/advanced/multi-task-learning.rst [hunk 14]
```diff
@@ -318,3 +318,129 @@ GraphStorm supports to run multi-task inference on :ref:`SageMaker<distributed-s
...
+the label field ``year`` will be ``edge_regression-paper_cite_paper-year``).
+The output format of task specific node embeddings follows
+the :ref:`GraphStorm saved node embeddings format<gs-out-embs>`.
+The ``save_embed_path`` in multi-task learning will look like following:
+
+.. code-block:: bash
+
+    emb_dir/
+        ntype0/
+            embed_nids-00000.pt
```
Comment (zhjwy9343) [inline docs/source/advanced/multi-task-learning.rst:351]: Could you check the latest output format of embedding file names. When run the latest code for node classification embedding generation, the output file name looks like:

`embed-00000_00000.parquet`

File: docs/source/advanced/multi-task-learning.rst [hunk 15]
```diff
@@ -318,3 +318,129 @@ GraphStorm supports to run multi-task inference on :ref:`SageMaker<distributed-s
...
+            embed_nids-00001.pt
+            ...
+            embed-00000.pt
+            embed-00001.pt
+            ...
+        emb_info.json
+        link_prediction-paper_cite_paper/
+            ntype0/
+                embed_nids-00000.pt
```
Comment (zhjwy9343) [inline docs/source/advanced/multi-task-learning.rst:367]: ditto

File: docs/source/advanced/multi-task-learning.rst [hunk 16]
```diff
@@ -318,3 +318,129 @@ GraphStorm supports to run multi-task inference on :ref:`SageMaker<distributed-s
...
+named with the
+corresponding task id. The output format of task specific prediction results
+follows the :ref:`GraphStorm saved prediction result format<gs-out-predictions>`.
+The ``save_prediction_path`` in multi-task learning will look like following:
+
+.. code-block:: bash
+
+    prediction_dir/
+        edge_regression-paper_cite_paper-year/
+            paper_cite_paper/
```
Comment (zhjwy9343) [inline docs/source/advanced/multi-task-learning.rst:421]: could you check if this is the latest format. I run the latest code for node classification, the prediction file name looks like:

`pred.predict-00000_00000.parquet`

File: docs/source/cli/model-training-inference/output.rst [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+of each node type are saved separately under different sub-directories named with
+the corresponding node types. GraphStorm will also save an ``emb_info.json`` file,
+which contains all the metadata for the saved node embeddings. The ``save_embed_path``
+will look like following:
+
+.. code-block:: bash
+
+    emb_dir/
+        ntype0/
+            embed_nids-00000.pt
```
Comment (zhjwy9343) [inline docs/source/cli/model-training-inference/output.rst:21]: ditto

File: docs/source/cli/model-training-inference/output.rst [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+            embed_nids-00001.pt
+            ...
+            embed-00000.pt
+            embed-00001.pt
+            ...
+        ...
+        emb_info.json
+
+The ``embed_nids-*`` files store the integer node IDs of each node embedding and
```
Comment (zhjwy9343) [inline docs/source/cli/model-training-inference/output.rst:37]: If the file name and file format change, need to revise these two sentences two reflect these changes.

File: docs/source/cli/model-training-inference/output.rst [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+            predict-00000.pt
+            predict-00001.pt
+            ...
+            predict_nids-00000.pt
+            predict_nids-00001.pt
+            ...
+        ...
+        result_info.json
+
+The ``predict_nids-*`` files store the integer node IDs of each prediction result and
```
Comment (zhjwy9343) [inline docs/source/cli/model-training-inference/output.rst:92]: ditto

File: docs/source/cli/model-training-inference/output.rst [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+        ...
+        emb_info.json
+
+The ``embed_nids-*`` files store the integer node IDs of each node embedding and
+the ``embed-*`` files store the corresponding node embeddings.
+The content of ``embed_nids-*`` files and ``embed-*`` files looks like:
+
+.. code-block::
+
+    embed_nids-00000.pt  |   embed-00000.pt
```
Comment (zhjwy9343) [inline docs/source/cli/model-training-inference/output.rst:43]: Our latest code's outputs look like:
```

>>> import pandas as pd
>>> preds = pd.read_parquet('./pred.predict-00000_00000.parquet')
>>> preds.head(10)
   pred  nid
0     0   p4
1     0  p11
2     0  p21
3     1  p38
4     0  p52
5     4  p53
6    11  p61
7     0  p66
8     4  p83
9    11  p92
```

File: docs/source/cli/model-training-inference/output.rst [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+        ...
+        result_info.json
+
+The ``predict_nids-*`` files store the integer node IDs of each prediction result and
+the ``predict-*`` files store the corresponding prediction results.
+The content of ``predict_nids-*`` files and ``predict-*`` files looks like:
+
+.. code-block::
+
+    predict_nids-00000.pt  |   predict.pt
```
Comment (zhjwy9343) [inline docs/source/cli/model-training-inference/output.rst:98]: Our latest code's outputs look like:

```
>>> import pandas as pd
>>> preds = pd.read_parquet('./pred.predict-00000_00000.parquet')
>>> preds.head(10)
   pred  nid
0     0   p4
1     0  p11
2     0  p21
3     1  p38
4     0  p52
5     4  p53
6    11  p61
7     0  p66
8     4  p83
9    11  p92
```

File: docs/source/cli/model-training-inference/output.rst [hunk 22]
```diff
@@ -0,0 +1,169 @@
...
+    Graph Node ID          |   Prediction results
+    10                     |   0.112
+    1                      |   0.872
+    23                     |   0.472
+    ...
+
+The ``result_info.json`` stores three informations:
+  * ``format``: The format of the saved prediction results. By default, it is ``pytorch``.
+  * ``emb_name``: A list of node types that have node prediction results saved. For example: ["ntype0", "ntype1"]
+  * ``world_size``: The number of chunks (files) into which the prediction results of a particular node type are divided. For instance, if world_size is set to 8, there will be 8 files for each set of prediction results."
```
Comment (zhjwy9343) [inline docs/source/cli/model-training-inference/output.rst:109]:
```suggestion
  * ``world_size``: The number of chunks (files) into which the prediction results of a particular node type are divided. For instance, if world_size is set to 8, there will be 8 files for each node type's prediction results."
```

File: docs/source/cli/model-training-inference/output.rst [hunk 23]
```diff
@@ -0,0 +1,169 @@
...
+
+
+For edge prediction tasks, the prediction results are saved per edge type.
+The sub-directory for an edge type is named as ``<src_ntype>_<relation_type>_<dst_ntype>``.
+For instance, given an edge type ``("movie","rated-by","user")``, the corresponding
+sub-directory is named as ``movie_rated-by_user``.
+GraphStorm will also save an ``result_info.json`` file, which contains all
+the metadata for the saved prediction results. The ``save_prediction_path``
```
Comment (zhjwy9343) [inline docs/source/cli/model-training-inference/output.rst:117]:
```suggestion
the metadata for the saved prediction results. The contents of the ``save_prediction_path``
```

File: docs/source/cli/model-training-inference/output.rst [hunk 24]
```diff
@@ -0,0 +1,169 @@
...
+will look like following:
+
+.. code-block:: bash
+
+    prediction_dir/
+        etype0/
+            predict-00000.pt
```
Comment (zhjwy9343) [inline docs/source/cli/model-training-inference/output.rst:124]: ditto

File: docs/source/cli/model-training-inference/output.rst [hunk 25]
```diff
@@ -0,0 +1,169 @@
...
+            src_nids-00000.pt
+            src_nids-00001.pt
+            ...
+            dst_nids-00000.pt
+            dst_nids-00001.pt
+            ...
+        ...
+        result_info.json
+
+The ``src_nids-*`` and ``dst_nids-*`` files contain the integer node IDs for
```
Comment (zhjwy9343) [inline docs/source/cli/model-training-inference/output.rst:146]: ditto

File: docs/source/cli/model-training-inference/output.rst [hunk 26]
```diff
@@ -0,0 +1,169 @@
...
+the source and destination nodes of each prediction, respectively.
+The ``predict-*`` files store the corresponding prediction results.
+The content of ``src_nids-*``, ``dst_nids-*`` and ``predict-*`` files looks like:
+
+.. code-block::
+
+    src_nids-00000.pt   |   dst_nids-00000.pt   |   predict.pt
```
Comment (zhjwy9343) [inline docs/source/cli/model-training-inference/output.rst:153]: ditto

File: docs/source/cli/model-training-inference/output.rst [hunk 27]
```diff
@@ -0,0 +1,169 @@
...
+    Source Node ID      |   Destination Node ID |   Prediction results
+    10                  |   12                  |   0.112
+    1                   |   20                  |   0.872
+    23                  |   3                   |   0.472
+    ...
+
+The ``result_info.json`` stores three informations:
+  * ``format``: The format of the saved prediction results. By default, it is ``pytorch``.
+  * ``etypes``: A list of edge types that have edge prediction results saved. For example: [("movie","rated-by","user"), ("user","watched","movie")]
+  * ``world_size``: The number of chunks (files) into which the prediction results of a particular edge type are divided. For instance, if world_size is set to 8, there will be 8 files for each set of prediction results."
```
Comment (zhjwy9343) [inline docs/source/cli/model-training-inference/output.rst:164]:
```suggestion
  * ``world_size``: The number of chunks (files) into which the prediction results of a particular edge type are divided. For instance, if world_size is set to 8, there will be 8 files for each node type's prediction results."
```

File: docs/source/cli/model-training-inference/output.rst [hunk 28]
```diff
@@ -0,0 +1,169 @@
...
+
+**Note: The built-in GraphStorm inference pipeline
+(launched by GraphStorm CLI) will process the saved prediction results
```
Comment (zhjwy9343) [inline docs/source/cli/model-training-inference/output.rst:167]:
```suggestion
(launched by GraphStorm CLIs) will process the saved prediction results
```

File: docs/source/cli/model-training-inference/output.rst [hunk 29]
```diff
@@ -0,0 +1,169 @@
...
+to convert the integer node ids into the raw node ids, which are usually string node ids.**
```
Comment (zhjwy9343) [inline docs/source/cli/model-training-inference/output.rst:168]:
```suggestion
to convert the integer node IDs into the raw node IDs, which are usually string IDs.**
```

File: docs/source/advanced/multi-task-learning.rst (hunk 14 above)
Comment (classicsong) [reply docs/source/advanced/multi-task-learning.rst:351]: This is the output of remap script.

File: docs/source/cli/model-training-inference/output.rst (hunk 18 above)
Comment (classicsong) [reply docs/source/cli/model-training-inference/output.rst:37]: Sure. But I don't think we will change it in the future.

File: docs/source/cli/model-training-inference/output.rst [hunk 30]
```diff
@@ -0,0 +1,214 @@
...
+        result_info.json
+
+The ``src_nids-*`` and ``dst_nids-*`` files contain the integer node IDs for
+the source and destination nodes of each prediction, respectively.
+The ``predict-*`` files store the corresponding prediction results.
+The content of ``src_nids-*``, ``dst_nids-*`` and ``predict-*`` files looks like:
+
+.. code-block::
+
+    src_nids-00000.pt   |   dst_nids-00000.pt   |   predict.pt
```
Comment (zhjwy9343) [inline docs/source/cli/model-training-inference/output.rst:196]:
```suggestion
    src_nids-00000.pt   |   dst_nids-00000.pt   |   predict-00000.pt
```

File: docs/source/cli/model-training-inference/output.rst [hunk 31]
```diff
@@ -0,0 +1,214 @@
...
+                        |
+    Source Node ID      |   Destination Node ID |   Prediction results
+    10                  |   12                  |   0.112
+    1                   |   20                  |   0.872
+    23                  |   3                   |   0.472
+    ...
+
+The ``result_info.json`` stores three informations:
```
Comment (zhjwy9343) [inline docs/source/cli/model-training-inference/output.rst:204]:
```suggestion
The ``result_info.json`` stores three types of informations:
```

File: docs/source/advanced/multi-task-learning.rst [hunk 32]
```diff
@@ -318,3 +318,129 @@ GraphStorm supports to run multi-task inference on :ref:`SageMaker<distributed-s
...
+                embed-00001.pt
+                ...
+            emb_info.json
+
+In the above example both the link prediction and edge regression tasks
+apply task specific normalization on node embeddings.
+
+**Note: The built-in GraphStorm training or inference pipeline
+(launched by GraphStorm CLIs) will process each saved node embeddings
+to convert the integer node IDs into the raw node IDs, which are usually string node IDs..**
```
Comment (zhjwy9343) [inline docs/source/advanced/multi-task-learning.rst:403]:
```suggestion
to convert the integer node IDs into the raw node IDs, which are usually string node IDs.**
```

File: docs/source/advanced/multi-task-learning.rst [hunk 33]
```diff
@@ -318,3 +318,129 @@ GraphStorm supports to run multi-task inference on :ref:`SageMaker<distributed-s
...
+~~~~~~~~~~~~~~~~~~~~~~
+When ``save_embed_path`` is provided in the training configuration or the inference configuration,
+GraphStorm will save the node embeddings in the corresponding path.
+In multi-task learning, by default, GraphStorm will save the node embeddings
+produced by the GNN layer for every node type under the path specified by
+``save_embed_path``. The output format follows the :ref:`GraphStorm saved node embeddings
+format<gs-out-embs>`. Meanwhile, in multi-task learning, certain tasks might apply
+task specific normalization to node embeddings. For instance, a link prediction
+task might apply l2 normalization on each node embeddings. In certain cases, GraphStorm
+will also save the normalized node embeddings under ``save_embed_path``.
```
Comment (zhjwy9343) [inline docs/source/advanced/multi-task-learning.rst:334]:
```suggestion
will also save the normalized node embeddings under the ``save_embed_path``.
```

File: docs/source/advanced/multi-task-learning.rst [hunk 34]
```diff
@@ -318,3 +318,129 @@ GraphStorm supports to run multi-task inference on :ref:`SageMaker<distributed-s
...
+named with the corresponding task id. (A task id is formated as ``<task_type>-<ntype/etype(s)>-<label>``.
+For instance, the task id of a node classification task on the node type ``paper`` with the
+label field ``venue`` will be ``node_classification-paper-venue``. As another example,
+the task id of a link prediction task on the edge type ``(paper, cite, paper)`` will be
+``link_prediction-paper_cite_paper``
+and the task id of a edge regression task on the edge type ``(paper, cite, paper)`` with
+the label field ``year`` will be ``edge_regression-paper_cite_paper-year``).
+The output format of task specific node embeddings follows
+the :ref:`GraphStorm saved node embeddings format<gs-out-embs>`.
+The contents of  the ``save_embed_path`` in multi-task learning will look like following:
```
Comment (zhjwy9343) [inline docs/source/advanced/multi-task-learning.rst:345]:
```suggestion
The contents of the ``save_embed_path`` in multi-task learning will look like following:
```

**Review by zhjwy9343 — APPROVED** [review]
LGTM

**Pull request merged**